text/0003-hash-indexed-object-store.md

```diff
@@ -0,0 +1,624 @@
+
+### DAG/Consensus
+
+There are two ways of embedding consensus-related DAG data into the system, either the object format accompanies every single type of supported consensus with an enum variant, or a generic variant is used.
+
+The first alternative is not as a bad as it sounds; while it does constitute a leaky abstraction, it improves the general robustness of the system, which will know how to reject invalid node types and can handle them "properly" inside its storage.
+
+The second alternative is adding a generic
+
+```rust
+enum Object {
+    // ...
+    DagNode<Box<dyn DagNode>>
+}
+```
+
+variant and implementing variations of these inside the consensus component. This will result in a trait and the consensus level having to do a bit of serialization/deserialization, but does not leak the associated discriminator of the enum into the serialized representation.
+
+## Drawbacks
+
+[drawbacks]: #drawbacks
+
+The biggest issue is that changes might affect execution engine performance negatively. This is still open to discovery, see [unresolved questions](#unresolved-questions).
+
+There is a considerable effort required to update all components and the change might not be easy to implement incrementally.
+
+The whole change to the execution engine may be prohibitive because of external dependencies, i.e. consumers of said execution engine that are not our node. This _might_ be worked around with a trait that abstracts the coupling between `Object` and components away.
+
+## Rationale and alternatives
+
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+### Hashing without serialization
+
+One alternative is to not hash serialized contents of objects, but have every object provide hashing functionality itself. This has the advantage of being a little cheaper should an object in memory need to be hashed without being serialized; however this is likely a rare occurence. It also makes the objects independent of the serialization method chosen and might reduce changes needed inside the execution engine.
+
+The drawback is that hashing now has to be implemented for each object on its own, but more importantly it is possible to accidentally introduce security issues by forgetting to include a field in a hash after changing it. Any occurence of such a bug is likely to result in a security issue, as it allows crafting two different specimen of an object with the same hash.
+
+This also removes easy verification of objects by outside sources.
+
+### Keeping components generic
+
+We can keep the fetcher and gossiper generic to reuse more of the existing code. Under this model we will reduce the number of gossipers to two (e.g. network and objects) or three (network, objects, consensus). Moving the logic for "validating" (checking for completeness) of an object is also not necessarily moved into fetcher this way.
```

Comment:
Retaining the generic versions and simply providing concrete impls for Object might shave a little time off the implementation. Seems pragmatic to me.

Reply:
I suspect (hope) the changes there would be small either way.